File: app/src/main/java/com/azizbekian/spyur/misc/InfiniteScrollListener.java
package com.azizbekian.spyur.misc;

import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import com.azizbekian.spyur.listener.IRecyclerLoadingListener;

/**
 * A scroll listener for {@link RecyclerView} to load more items as you approach the end.
 * <p>
 * Adapted from <a href="https://gist.github.com/ssinss/e06f12ef66c51252563e">WoongBi Kim</a>
 */
public abstract class InfiniteScrollListener extends RecyclerView.OnScrollListener {

    /**
     * The minimum number of items remaining before some action should be performed.
     */
    private static final int VISIBLE_THRESHOLD = 5;

    private final LinearLayoutManager layoutManager;
    private final IRecyclerLoadingListener isLoading;
    private final int visibleThreshold;

    public InfiniteScrollListener(@NonNull LinearLayoutManager layoutManager,
                                  @NonNull IRecyclerLoadingListener isLoading,
                                  int visibleThreshold) {

        this.layoutManager = layoutManager;
        this.isLoading = isLoading;
        this.visibleThreshold = visibleThreshold;
    }

    @SuppressWarnings("unused")
    public InfiniteScrollListener(@NonNull LinearLayoutManager layoutManager,
                                  @NonNull IRecyclerLoadingListener isLoading) {

        this(layoutManager, isLoading, VISIBLE_THRESHOLD);
    }

    @Override
    public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
        // bail out if scrolling upward or already loading data
        if (dy < 0 || isLoading.isDataLoading()) return;

        final int visibleItemCount = recyclerView.getChildCount();
        final int totalItemCount = layoutManager.getItemCount();
        final int firstVisibleItem = layoutManager.findFirstVisibleItemPosition();

        if ((totalItemCount - visibleItemCount) <= (firstVisibleItem + visibleThreshold)) {
            onLoadMore();
        }
    }

    /**
     * Is called, when the {@code threshold} has been passed, meaning that a new data should be
     * loaded.
     */
    public abstract void onLoadMore();

}

File: app/src/main/kotlin/com/incentive/yellowpages/misc/InfiniteScrollListener.kt
package com.incentive.yellowpages.misc

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView

/**
 * A scroll listener for [RecyclerView] to load more items as you approach the end.
 *
 *
 * Adapted from [WoongBi Kim](https://gist.github.com/ssinss/e06f12ef66c51252563e)
 */
abstract class InfiniteScrollListener(private val layoutManager: LinearLayoutManager,
                                      private val loadMoreListener: LoadMoreListener,
                                      private val visibleThreshold: Int) : RecyclerView.OnScrollListener() {

    override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
        // bail out if scrolling upward or already loading data
        if (dy < 0 || loadMoreListener.isDataLoading()) return

        val visibleItemCount = recyclerView!!.childCount
        val totalItemCount = layoutManager.itemCount
        val firstVisibleItem = layoutManager.findFirstVisibleItemPosition()

        if (totalItemCount - visibleItemCount <= firstVisibleItem + visibleThreshold) {
            onLoadMore()
        }
    }

    /**
     * Is called, when the `threshold` has been passed, meaning that a new data should be
     * loaded.
     */
    abstract fun onLoadMore()

}
